# AA_U2_T1_Numpy_GuerraPaul.ipynb

**PROGRAMACIÓN 2**  
**Actividad Autónoma 3**  
**Nombre de la actividad:** Análisis de consumo energético con NumPy  

**Nombre:** Paúl Andrés Guerra Vicuña  
**Fecha:** 24/05/2026  
**Carrera:** Ingeniería en Ciencias de Datos e Inteligencia Artificial  
**Periodo Académico:** 2026-1S  
**Semestre:** Segundo “A”

#Actividades que debe desarrollar el estudiante

#1.	Creación del arreglo NumPy

Crear un arreglo bidimensional llamado consumo con los datos de la tabla.




In [1]:
import numpy as np

# Datos de consumo (filas = dispositivos, columnas = días)
consumo = np.array([
    [12.5, 13.0, 12.8, 14.1, 13.7, 6.2, 4.8],   # Computadoras
    [6.5, 6.8, 7.0, 6.9, 7.2, 3.1, 2.5],       # Laptops
    [18.0, 18.2, 18.1, 18.3, 18.4, 18.0, 17.9], # Servidor
    [4.2, 4.1, 4.3, 4.4, 4.2, 4.0, 3.9],       # Router/Red
    [9.0, 9.5, 9.2, 9.8, 10.0, 3.5, 2.0]       # Iluminación
])

print("Matriz de consumo:\n", consumo)
print("\nNúmero de dimensiones:", consumo.ndim)
print("Forma del arreglo:", consumo.shape)
print("Número total de elementos:", consumo.size)
print("Tipo de dato:", consumo.dtype)

Matriz de consumo:
 [[12.5 13.  12.8 14.1 13.7  6.2  4.8]
 [ 6.5  6.8  7.   6.9  7.2  3.1  2.5]
 [18.  18.2 18.1 18.3 18.4 18.  17.9]
 [ 4.2  4.1  4.3  4.4  4.2  4.   3.9]
 [ 9.   9.5  9.2  9.8 10.   3.5  2. ]]

Número de dimensiones: 2
Forma del arreglo: (5, 7)
Número total de elementos: 35
Tipo de dato: float64


Luego mostrar:

1.	La matriz completa.
2.	El número de dimensiones.
3.	La forma del arreglo.
4.	El número total de elementos.
5.	El tipo de dato almacenado.


#2.Análisis general del consumo

Usando funciones de NumPy, calcular:

 1.	Consumo total semanal de todos los dispositivos.
 2.	Consumo promedio general.

 3.	Consumo máximo registrado.
 4.	Consumo mínimo registrado.


In [2]:
total_semanal = np.sum(consumo)
promedio_general = np.mean(consumo)
max_consumo = np.max(consumo)
min_consumo = np.min(consumo)

print("Consumo total semanal:", total_semanal, "kWh")
print("Consumo promedio general:", round(promedio_general, 2), "kWh")
print("Consumo máximo registrado:", max_consumo, "kWh")
print("Consumo mínimo registrado:", min_consumo, "kWh")

Consumo total semanal: 326.1 kWh
Consumo promedio general: 9.32 kWh
Consumo máximo registrado: 18.4 kWh
Consumo mínimo registrado: 2.0 kWh


Después, escribir una interpretación breve sobre el comportamiento general del consumo energético.

**Interpretación breve:**

En general, la oficina consumió 326.1 kWh durante la semana, el consumo promedio por registro fue de 9.32 kWh pero tambien se observa que hay una gran diferencia entre los días normales y el fin de semana ya que durante la semana se consume mucho más energía, mientras que sábado y domingo baja considerablemente. Esto es normal porque en los días laborables se usan más los equipos.

#3.Análisis por dispositivo usando axis

Calcular el consumo total semanal por cada dispositivo.




In [3]:
total_por_dispositivo = np.sum(consumo, axis=1)
dispositivos = ["Computadoras", "Laptops", "Servidor", "Red", "Iluminación"]

for disp, total in zip(dispositivos, total_por_dispositivo):
    print(f"{disp}: {total:.1f} kWh")

# Desviación estándar para medir constancia
std_por_dispositivo = np.std(consumo, axis=1)
for disp, std in zip(dispositivos, std_por_dispositivo):
    print(f"{disp} - Std: {std:.2f}")

Computadoras: 77.1 kWh
Laptops: 40.0 kWh
Servidor: 126.9 kWh
Red: 29.1 kWh
Iluminación: 53.0 kWh
Computadoras - Std: 3.54
Laptops - Std: 1.86
Servidor - Std: 0.17
Red - Std: 0.16
Iluminación - Std: 3.09


**Luego responder:**

1. ¿Qué dispositivo consumió más energía durante la semana?  
   **El Servidor local**, con 126.9 kWh.

2. ¿Qué dispositivo consumió menos energía durante la semana?  
   **El Router y equipos de red**, con solo 29.1 kWh.

3. ¿Qué dispositivo mantiene un consumo más constante?  
   **El Servidor local**, es el que tiene el consumo más estable durante todos los días.

4. ¿Qué conclusión se puede obtener sobre el servidor local?  
   El servidor consume mucha energía todos los días de forma constante porque probablemente está encendido las 24 horas. Es el dispositivo que más energía usa en la oficina.

#4.	Análisis por día usando axis

Calcular el consumo total por cada día de la semana.


In [4]:
total_por_dia = np.sum(consumo, axis=0)
dias = ["Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo"]

for dia, total in zip(dias, total_por_dia):
    print(f"{dia}: {total:.1f} kWh")

Lunes: 50.2 kWh
Martes: 51.6 kWh
Miércoles: 51.4 kWh
Jueves: 53.5 kWh
Viernes: 53.5 kWh
Sábado: 34.8 kWh
Domingo: 31.1 kWh


**Luego responder:
1. ¿Qué día tuvo mayor consumo energético?  
   **El Jueves y el Viernes**, ambos con 53.5 kWh.

2. ¿Qué día tuvo menor consumo energético?  
   **El Domingo**, con solo 31.1 kWh.

3. ¿Qué diferencia se observa entre días laborables y fin de semana?  
   Los días de lunes a viernes se consume mucho más energía (alrededor de 50-53 kWh por día), mientras que sábado y domingo baja bastante (34.8 y 31.1 kWh). Esto se debe a que en los días laborales se usan más las computadoras, laptops e iluminación.

4. ¿Qué recomendación podría darse para optimizar el consumo?  
   Apagar las computadoras, laptops y luces al final del día y los fines de semana. También se podría revisar si es posible mejorar la eficiencia del servidor o apagarlo parcialmente los fines de semana.

#5.	**Transformación del arreglo**

 Transformar la matriz original en un arreglo unidimensional.

In [5]:
consumo_flat = consumo.flatten()

print("Arreglo unidimensional:\n", consumo_flat)
print("\nCantidad de elementos:", len(consumo_flat))
print("Promedio desde arreglo plano:", np.mean(consumo_flat))

Arreglo unidimensional:
 [12.5 13.  12.8 14.1 13.7  6.2  4.8  6.5  6.8  7.   6.9  7.2  3.1  2.5
 18.  18.2 18.1 18.3 18.4 18.  17.9  4.2  4.1  4.3  4.4  4.2  4.   3.9
  9.   9.5  9.2  9.8 10.   3.5  2. ]

Cantidad de elementos: 35
Promedio desde arreglo plano: 9.317142857142859


Luego:

1.	Mostrar el arreglo transformado.

Se muestra el arreglo unidimensional con todos los valores.

2.	Verificar cuántos datos contiene.

Contiene 35 datos.

3.	Calcular nuevamente el promedio general.

El promedio general es 9.32 kWh.

4.	Comparar este promedio con el obtenido desde la matriz original.

El promedio es exactamente el mismo que el calculado en la matriz original.

5.	Explicar por qué ambos resultados coinciden.

Ambos resultados coinciden porque al aplanar el arreglo solo se cambia la forma, pero los números siguen siendo los mismos.


# 6.	Mini análisis con condición
Crear una condición para identificar los consumos mayores a 15 kWh.

In [6]:
mayores_15 = consumo > 15
registros_altos = np.sum(mayores_15)

print("Registros > 15 kWh:", registros_altos)
print("Posiciones:", np.where(mayores_15))

Registros > 15 kWh: 7
Posiciones: (array([2, 2, 2, 2, 2, 2, 2]), array([0, 1, 2, 3, 4, 5, 6]))


**Luego responder:**

1. ¿Cuántos registros superan los 15 kWh?  
   **7 registros**.

2. ¿A qué dispositivo pertenecen principalmente esos valores?  
   **Todos pertenecen al Servidor local**.

3. ¿Qué indica esto sobre el patrón de consumo?  
   Indica que el servidor es el que más energía consume y lo hace de forma continua todos los días. Es importante revisar su consumo porque representa una parte muy grande del total de energía usada en la oficina.

#7.	Reflexión final

Redactar una conclusión de entre 8 y 10 líneas donde explique:

1.	Por qué NumPy es útil para analizar datos numéricos.
2.	Qué ventaja tiene trabajar con arreglos en lugar de listas tradicionales.
3.	Cómo ayuda axis a interpretar datos organizados en filas y columnas.
4.	En qué otros problemas reales podría aplicarse este tipo de análisis.


**Reflexión final:**

En mi opinion NumPy es muy útil para analizar datos numéricos porque permite hacer cálculos rápidos y sencillos con grandes cantidades de números. En este caso nos ayudó a organizar el consumo de energía y sacar totales, promedios y máximos fácilmente.

La ventaja de trabajar con arreglos de NumPy en lugar de listas normales de Python es que es mucho más rápido y permite hacer operaciones en todo el arreglo al mismo tiempo sin necesidad de usar bucles.

El parámetro axis es muy práctico porque nos permite analizar por filas (por dispositivo) o por columnas (por día) sin complicarnos. Gracias a esto pudimos entender mejor el comportamiento de cada equipo y de cada día de la semana.

Este tipo de análisis se puede aplicar en muchos problemas reales como: controlar el consumo de electricidad en una casa o edificio, analizar el rendimiento de servidores en una empresa, llevar el control de gastos, evaluar el uso de maquinaria en una fábrica, entre otros.

